File: backend/models/telemetry/telemetry.go
package telemetry

import (
	"fmt"
	"missioncontrol/backend/models/command"
	"strings"
)

type Telemetry struct {
	PacketLoss     int                           `json:"packetLoss"`
	Latency        float64                       `json:"latency"`
	DataRate       int                           `json:"dataRate"`
	Status         *command.ParsedStatusResponse `json:"status"`
	CommandLog     string                        `json:"commandLog"`
	AvailablePorts []string                      `json:"availablePorts"`
	CurrentPort    string                        `json:"currentPort"`
}

func (t *Telemetry) ToGrafanaString() string {
	boolToInt := func(b bool) int {
		if b {
			return 1
		}
		return 0
	}

	header := "telemetry,source=backend"

	fields := []string{
		// Filling
		fmt.Sprintf("N2O\\ Tank\\ Pressure=%f", t.Status.HydraFS.N2O.Pressure),
		// TODO: Loadcell tanque
		fmt.Sprintf("N2O\\ Bottle\\ Weight=%f", t.Status.LiftFS.N2OLoadcell),
		fmt.Sprintf("Vent\\ Valve=%d", t.Status.HydraUF.VentValve.Binarize()),
		fmt.Sprintf("N2O\\ Fill\\ Valve=%d", t.Status.HydraFS.N2O.FillValve.Binarize()),
		fmt.Sprintf("N2\\ Fill\\ Valve=%d", t.Status.HydraFS.N2.FillValve.Binarize()),

		// Launch
		fmt.Sprintf("Lift\\ Thrust\\ 1\\ Weight=%f", t.Status.LiftR.Loadcell1),
		fmt.Sprintf("Lift\\ Thrust\\ 2\\ Weight=%f", t.Status.LiftR.Loadcell2),
		fmt.Sprintf("Lift\\ Thrust\\ 3\\ Weight=%f", t.Status.LiftR.Loadcell3),
		fmt.Sprintf("Main\\ Valve=%d", t.Status.HydraLF.MainValve.Binarize()),
		fmt.Sprintf("Main\\ Parachute\\ Ematch=%d", boolToInt(t.Status.Elytra.MainChuteEmatch)),
		fmt.Sprintf("Drogue\\ Parachute\\ Ematch=%d", boolToInt(t.Status.Elytra.DrogueChuteEmatch)),
		fmt.Sprintf("Chamber\\ Pressure=%f", t.Status.HydraLF.ChamberPressure),
		fmt.Sprintf("Altitude=%f", t.Status.Navigator.GPS.Altitude),
		fmt.Sprintf("Velocity\\ Z=%f", t.Status.Navigator.Kalman.VelocityZ),
		fmt.Sprintf("Acceleration\\ Z=%f", t.Status.Navigator.Kalman.AccelerationZ),

		// Recovery
	}

	return fmt.Sprintf("%s %s", header, strings.Join(fields, ","))
}

File: frontend/src/utils/models/telemetry.ts
import type { MissionState } from './missionstate';
import type { ValveState } from './valve';

export type Telemetry = {
  packetLoss: number;
  latency: number;
  dataRate: number;
  status: {
    obc: OBC;
    hydraUf: HydraUF;
    hydraLf: HydraLF;
    hydraFs: HydraFS;
    navigator: Navigator;
    elytra: Elytra;
    liftR: LiftR;
    liftFs: LiftFS;
  };
  commandLog: string;
  availablePorts: string[];
  currentPort: string;
};

type OBC = {
  state: MissionState;
  sdLogging: boolean;
};

type HydraUF = {
  ventValve: ValveState;
  pressurizingValve: ValveState;
  probeThermo1: number;
  probeThermo2: number;
  probeThermo3: number;
};

type HydraLF = {
  mainValve: ValveState;
  abortValve: ValveState;
  probeThermo1: number;
  probeThermo2: number;
  chamberTemperature: number;
  tankPressure: number;
  chamberPressure: number;
};

type HydraFS = {
  n2: N2;
  n2o: N2O;
  quickDc: QuickDC;
};

type N2 = {
  fillValve: ValveState;
  purgeValve: ValveState;
  pressure: number;
};

type N2O = {
  fillValve: ValveState;
  purgeValve: ValveState;
  pressure: number;
  temperature: number;
};

type QuickDC = {
  n2Valve: ValveState;
  n2oValve: ValveState;
  pressure: number;
};

type Navigator = {
  gps: GPS;
  kalman: Kalman;
  imu: IMU;
  barometerAlt1: number;
  barometerAlt2: number;
};

type GPS = {
  latitude: number;
  longitude: number;
  satCount: number;
  altitude: number;
  horizontalVelocity: number;
};

type Kalman = {
  velocityZ: number;
  accelerationZ: number;
  altitude: number;
  maxAltitude: number;
  q1: number;
  q2: number;
  q3: number;
  q4: number;
};

type IMU = {
  accelX: number;
  accelY: number;
  accelZ: number;
  gyroX: number;
  gyroY: number;
  gyroZ: number;
  magX: number;
  magY: number;
  magZ: number;
};

type Elytra = {
  mainChuteEmatch: boolean;
  drogueChuteEmatch: boolean;
};

type LiftR = {
  loadcell1: number;
  loadcell2: number;
  loadcell3: number;
  mainEmatch: boolean;
};

type LiftFS = {
  n2oLoadcell: number;
};
